### PR TITLE
refactor(go): convert to class-based datasource

### PR DESCRIPTION
**Pull request by JamieMagee**
## Changes:

Converts the `go` datasource to be class-based

TODO:

- [x] ~Run on a real repo~ https://github.com/renovate-reproductions/12985/pull/1

## Context:

#8647

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/datasource/go/index.ts [hunk 1]
```diff
@@ -1,17 +1,71 @@
...
+    { lookupName }: Partial<DigestConfig>,
+    value?: string
+  ): Promise<string | null> {
+    const source = await BaseGoDatasource.getDatasource(lookupName);
+    if (!source) {
+      return null;
+    }
+
+    // ignore v0.0.0- pseudo versions that are used Go Modules - look up default branch instead
+    const tag = value && !value.startsWith('v0.0.0-2') ? value : undefined;
```
Comment (viceice) [inline lib/datasource/go/index.ts:56]: `'v0.0.0-2'`: shouldn't this be `'v0.0.0-'`, as the comment says?

//cc @rarkins

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/datasource/go/index.ts (hunk 1 above)
Comment (rarkins) [reply lib/datasource/go/index.ts:56]: I guess it's looking for years which follow it, e.g. `v0.0.0-2021.....`

Comment (JamieMagee) [reply lib/datasource/go/index.ts:57]: I'm not super familar with Go, but Rhys added it, so I'll defer to him. See https://github.com/renovatebot/renovate/pull/3267/files#diff-67dacc71c63f241e34f992308f1a4e1296198f914c6cb9a91f642d5faeb2e5cf

**Comment by JamieMagee** [comment]
Run on a simple repo https://github.com/renovate-reproductions/12985/pull/1

**Comment by rarkins** [comment]
I think it also needs a manual repo test that GOPROXY env still works as expected. I can try doing that

File: lib/datasource/go/base.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        return {
+          datasource: github.id,
+          lookupName: goSourceUrl
+            .replace('https://github.com/', '')
+            .replace(regEx(/\/$/), ''),
+          registryUrl: 'https://github.com',
+        };
+      }
+      const gitlabUrl =
+        BaseGoDatasource.gitlabHttpsRegExp.exec(goSourceUrl)?.groups
```
Comment (viceice) [inline lib/datasource/go/base.ts:85]: Please create local const for `BaseGoDatasource.gitlabHttpsRegExp.exec(goSourceUrl)`

**Review by viceice — APPROVED** [review]

**Review by rarkins — APPROVED** [review]

**Pull request merged**